### DISCUSSION TITLE
How to properly handle HTTP status codes in SendHttpRequest?

### DISCUSSION THREAD

**Discussion by Bergam64** [elsa-workflows/elsa-core#2691] [Q&A]
Hello,

I'm new to ELSA (I'm using 2.3) and I still haven't figured out what's the best way to handle the possible results of a `SendHttpRequest`.
Note: I don't use the designer to build my workflows, only builders.

I have a very common use case here:
- If `SendHttpRequest` returned a successful HTTP status code (like 200 or 201), I want to execute another part of my workflow.
- If it returned an error HTTP status code (like 400 or 500), I want that code **and** the response body to end up in the log (so I have a clue of what went wrong), and potentially I want to execute another part of my workflow.
- If another error occurs (e.g. the target server can't be reached), the workflow should fail.

Problems:
- In case of an error HTTP status code, only that code ends up in the log, but never the response body, even when `.WithReadContent(true)` has been specified on the `SendHttpRequest`. This makes it **very difficult** to identify the root cause of the failed HTTP request.
- At design time, only 2 outcomes are available for connections: "Done" and "Unsupported Status Code". There is no provision for:
  - A "Supported Status Code" outcome.
  - The "200" outcome (which corresponds to the default supported HTTP status code in `SendHttpRequest`).
  - The string representation of the HTTP status code(s) that you specify  with `.WithSupportedHttpCodes`.

Using the "Done" outcome is **not** acceptable because once connected to another activity, that other activity is **also** executed when `SendHttpRequest` has returned an error HTTP status code.

I've tried to use `.When("200")` to connect with HTTP status code 200, but it doesn't work as expected:
- At design time, since no outcome "200" exists on `SendHttpRequest`, **no connection is created** (and ELSA does **not** notify you of this problem, which is questionable).
- But after the workflow has been executed and `SendHttpRequest` has returned HTTP status code 200, then the connection is finally visible in the workflow instance. This also implies that, for a given workflow, its diagram and its instance diagrams may be significantly different in terms of connections, which is questionable.

I even tried to use an `If` activity to manually check the HTTP status code, but getting the `Response` property of the `SendHttpRequest` with `GetNamedActivityPropertyAsync` is impossible, as `If` has no provision for an async expression for the first parameter.

So I'm stuck. Please help!

One last request: the lack of XML comments in most ELSA code (which means that no documentation is displayed in VS with IntelliSense), makes the learning curve of ELSA very slow. Please kindly consider adding XML comments to the public surface of the ELSA APIs.

Thanks in advance.

**Comment by sfmskywalker** [comment]
@Bergam64 

Thanks for the great feedback. Some of the things you mention require improvements to be made, others may need some clarification. The comment on lack of XML comments is duly noted 👍🏻 

> In case of an error HTTP status code, only that code ends up in the log, but never the response body, even when .WithReadContent(true) has been specified on the SendHttpRequest. This makes it very difficult to identify the root cause of the failed HTTP request.

Although the response **does** get recorded in the log, it's currently not exposed through the UI. This is something we need to add.
Until then, the easiest way to figure out what went wrong is to look at the `Data` field of the workflow instance record.

For example, take the following workflow that makes an HTTP requests but with a missing field that is required by the API:

![send-http-request-check-response](https://user-images.githubusercontent.com/938393/153650330-05db3ae0-77c4-42ad-8046-26b45ea21210.gif)

> At design time, only 2 outcomes are available for connections: "Done" and "Unsupported Status Code". There is no provision for:
> A "Supported Status Code" outcome.
> The "200" outcome (which corresponds to the default supported HTTP status code in SendHttpRequest).
> The string representation of the HTTP status code(s) that you specify with .WithSupportedHttpCodes.

Maybe this is a bug in 2.3, but I don't see this issue when I try with the current version. I am able to configure multiple outcomes I wish to handle in the workflow, like this:

![send-http-request-expected-outcomes](https://user-images.githubusercontent.com/938393/153651114-fa42f871-9a55-45f9-9047-43bbe3dac535.gif)

With regards to the programmatic API however, I see your point and is a complete oversight in the extension method for the SendHttpRequest API. What we need is an additional parameter that lets you create connections from the activity to subsequent activities. E.g.:

```csharp
builder.SendHttpRequest(activity => activity
    .WithUrl(new Uri("https://reqres.in/api/login"))
    .WithMethod(HttpMethod.Post.Method)
    .WithContent("{\"email\": \"peter@klaven\"}")
    .WithReadContent(true)
    .WithSupportedHttpCodes(new[]{ 200, 400, 500 }),
    activity =>
    {
        activity.When("200").WriteLine("OK!");
        activity.When("400").WriteLine("Missing some fields!");
        activity.When("500").WriteLine("Server is sad :(");
    });
}
```

I just pushed an update that allows you to do the above. Once this build completes successfully, you can try it using the latest MyGet preview packages if you want. If you don't want to upgrade to 2.5.1 prereleases yet, you could try to following syntax instead:

```csharp
builder.Then<SendHttpRequest>(activity => activity
    .WithUrl(new Uri("https://reqres.in/api/login"))
    .WithMethod(HttpMethod.Post.Method)
    .WithContent("{\"email\": \"peter@klaven\"}")
    .WithReadContent(true)
    .WithSupportedHttpCodes(new[] { 200, 400 }), activity =>
{
    activity.When("200").WriteLine("OK!");
    activity.When("400").WriteLine("Missing some fields!");
    activity.When("500").WriteLine("Server is sad :(");
});
```

Which is almost the same, the difference being that, here I'm using `Then<SendHttpRequest>(...)` explicitly, rather than `SendHttpRequest(...)` (which is just an extension method that does the exact same thing).

**Reply by Bergam64** [reply]
Hello Sipke,

Thanks for your detailed reply and sorry for my giving you an update so late. I've been pretty busy lately.

I've tried your suggestion with `.Then<SendHttpRequest>` in ELSA 2.3, but that didn't fix anything.

Let me explain once more what exactly my issue is:
- All my workflows are built by code, using the builder API (never using the designer).
- When I connect a `SendHttpRequest` activity with another activity for HTTP status code 200 using `.When("200")`, this connection **never appears** in the diagram when I open the workflow in the Dashboard by clicking "Workflow Registry" an I click on my workflow in the "NAME" column (not in the "INSTANCES" column). Visually, in the workflow diagram, for each `SendHttpRequest` activity, I end up with a workflow part that is _disconnected_ from the main workflow part (the one attached to the `Start` activity), due to the missing connection for outcome `"200"`.
- In other words, the diagram of a workflow with a `SendHttpRequest` activities and its corresponding instance diagrams can be significantly different.
- Whether or not I call `.WithSupportedHttpCodes(new[] { 200 })` on the `SendHttpRequest` activity doesn't change anything.
- If I execute the workflow and the scenario produces HTTP status 200 on the `SendHttpRequest` activity, then **at this point** the connection for outcome `"200"` finally **appears** in the corresponding workflow **instance** diagram, but still not in the workflow diagram.
- I typically have several `SendHttpRequest` activities per workflow, so this issue makes it very difficult to verify the structure of complex workflows without first carefully testing them with all scenarios and checking the corresponding workflow instance diagrams.

From what I understand from ELSA 2.3 source code:
- At workflow build time, the only available outcomes are `"Done"` and `"Unsupported Status Code"`, whatever may have been configured with `.WithSupportedHttpCodes`. So any connection attempt with any other outcome (e.g. `"200"`) _silently_ fails.
- At `SendHttpRequest` activity execution time, the received HTTP status code (say, `200`) is turned to a string and added to the activity outcomes, which somehow magically resurrects the connection initially set up with the builder API.

Is my understanding correct?

Last Friday, I finally had the time to upgrade to more recent releases of ELSA, but to no avail. Here are the results:

- With ELSA 2.3: problem still present.
- After upgrading to 2.4.0.1: problem still present.
- Ditto after upgrading to 2.4.1, 2.4.2 and 2.4.2.1.
- After upgrading to 2.5.0, it's can't even tell, because **my workflows no longer appear in the Workflow Registry** in the Dashboard (the list is empty)! Do I need to do anything special to have my workflows recognized in 2.5.0?

I'm really stuck and I still have dozens more workflows to implement (which will all hit this issue), please help!

**Reply by sfmskywalker** [reply]
Hey man, I just checked and I see what you mean:

Here I have a simple workflow:

```csharp
public class SendHttpWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder.WithWorkflowDefinitionId("SendHttp")
            .WithVersion(1)
            .WithDescription("Send test http request and return response")
            .SendHttpRequest(setup => setup.WithUrl(new Uri("https://jsonplaceholder.typicode.com/todos/3"))
                .WithMethod("GET")
                .WithSupportedHttpCodes(new[]{200, 400, 500})
                .WithReadContent(true), sendHttp =>
            {
                sendHttp.When("200").WriteHttpResponse(HttpStatusCode.OK, "OK!", MediaTypeNames.Text.Plain);
                sendHttp.When("400").WriteHttpResponse(HttpStatusCode.BadRequest, "Bad Request!", MediaTypeNames.Text.Plain);
                sendHttp.When("500").WriteHttpResponse(HttpStatusCode.BadRequest, "Bad Request!", MediaTypeNames.Text.Plain);
            });
    }
}
```

And here is what it looks like in the workflow blueprint viewer:

![image](https://user-images.githubusercontent.com/938393/157220974-8e9d1a24-19fa-4d79-96ef-95460e808db7.png)

Not what I had expected, so I'll be fixing this shortly.

> After upgrading to 2.5.0, it's can't even tell, because my workflows no longer appear in the Workflow Registry in the Dashboard (the list is empty)! Do I need to do anything special to have my workflows recognized in 2.5.0?

This I can't reproduce. When I look at the workflow registry, I see my workflow:

![image](https://user-images.githubusercontent.com/938393/157222473-8c724718-38b2-46a3-838f-79653acddd53.png)


If you can provide me with steps to reproduce then I'll take a look at that as well.

**Reply by sfmskywalker** [reply]
I pushed a change (a9fe7bd5f25e888be9d97369c1ab13c16f949dfa) that fixes the designer issue. Once [this build](https://ci.appveyor.com/project/sfmskywalker/elsa/builds/42825274) completes, you should be able to try out the latest packages from MyGet. The fix will be part of 2.6.0 on NuGet which will hopefully ship later this month.

**Comment by Bergam64** [comment]
Hello!

Thanks for your quick reply. I'm really looking forward to having this issue fixed!

I've seen that the bug you've created for this has been added to the 2.5.1 milestone, but that milestone is no logner part of the [milestones page](https://github.com/elsa-workflows/elsa-core/milestones). The earliest next milestone listed there is now 2.6.0. Will 2.5.1 be in fact 2.6.0?

Regarding the issue with 2.5.0, here's what I've done:
- I've started my project with ELSA 2.3.0. I've created all my workflows with the builder API. I have zero workflow created by the designer.
- Then I've started incrementally upgrading to each version up to 2.5.0, as follows.
  - Before starting these upgrades, I've first deleted all by workflows instances using the Dashboard, to make sure the breaking change mentioned in ELSA 2.4.0 (about a new column in table `WorkflowInstances`) doesn't cause any issue. Then I've created a DB snapshot of my ELSA DB 2.3.0, just in case.
  - I've upgraded to ELSA 2.4.0.1 by just upgrading the ELSA NuGet packages and recompiling. No code change was required. At runtime, everything was look OK in the Dashboard, except that the `SendHttpRequest` issue mentioned above was still present.
  - I've done the same for ELSA 2.4.1, then for 2.4.2, then for 2.4.2.1. So far so good.
  - Then finally I've upgraded to ELSA 2.5.0 but at runtime, my workflows were gone from the registry in the Dashboard!

Note: I've finally reverted all ELSA NuGet packages back to 2.3.0 and restored my ELSA 2.3.0 DB using my DB snapshot. Then I've re-applied all upgrades up to 2.4.2.1. This is where I'm now. My workflows still appear in the registry.

I hope this is enough for you to repro that issue.

**Comment by sfmskywalker** [comment]
> I've seen that the bug you've created for this has been added to the 2.5.1 milestone, but that milestone is no logner part of the [milestones page](https://github.com/elsa-workflows/elsa-core/milestones). The earliest next milestone listed there is now 2.6.0. Will 2.5.1 be in fact 2.6.0?

That is indeed confusing! I renamed the 2.5.1 milestone to 2.6.0 - I should probably have just closed 2.5.1 instead and create 2.6.0 😄

**Comment by Bergam64** [comment]
Hello Sipke,

I've just attempted an upgrade from ELSA 2.4.2.1 to 2.6.0, but unfortunately the result is exactly the same as with my previous attempt with 2.5.0: **none of my code-built workflows appear anymore in the Workflow Registry in the Dashboard!** So once again, I had to roll back the upgrade.

This means that **I'm stuck with ELSA 2.4.2.1** and consequently **I can't get any bug fix** (especially [#2748](https://github.com/elsa-workflows/elsa-core/issues/2748), which was initiated by this discussion).

I have no clue what's specific to my situation and why you can't reproduce it with a code-built workflow created with a previous version (say, 2.4.2.1) and then all ELSA NuGet packages are upgraded to 2.6.0. A question you should ask yourself is  

_what in ELSA (especially 2.5.0 and later) can cause a code-built workflow never to appear in the Registry?_

**Please help!**

**Reply by sfmskywalker** [reply]
Hi @Bergam64, I’m truly sorry to hear about your struggles. Is there any way for you to provide a reproducible solution that I can look at? Unfortunately I do not currently have the bandwidth to try and reproduce this myself - but if I have a series of steps (or a broken solution demonstrating the issue) then I’m sure we can get this fixed.

**Comment by Bergam64** [comment]
Hello Sipke,

Sorry for replying so late, but as you may have guessed, I also don't have much bandwidth either.

Nevertheless, I have some good news : I have been finally able to identify the root cause of the problem!

Once again, I've upgraded my project from ELSA 2.4.2.1 to 2.5.0 and once again, the Workflow Registry was displaying an empty list. But the subtle thing I hadn't noticed so far in your last screenshot was the introduction of the new workflow provider dropdown menu in that page :
![image](https://user-images.githubusercontent.com/29800159/162217646-de62f962-877a-46f4-bc85-7475b825b28a.png)
which was **never** displayed in my case _until_ I had the idea of **force-refreshing** the page (in Chrome, use Ctrl+**Shift**+R instead of Ctrl+R), which has revealed both that dropdown menu **and my workflows**!

Please try to find a way to fix the issue, and also add a **BIG WARNING** in the release page of release 2.5.0 about this cache issue.

Hope this helps.

Best regards
